fix(config/validation): validate `allowedValues`

jamietanna opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes







To make sure that we provide faster feedback for users, we should
validate that for options that have an `allowedValues` set, we validate
the option according to the value.

We also need to:

- handle a single option value
- handle an array of option values
- handle when a default (`null`) option is used, as that won't match
  `allowedValues`
- handle when a `supportsTemplating` config option is used alongside
  `allowedValues`
- add a special case to handle `packageRule.versioning` differently, as
  it may have a `regex:` value which isn't technically in the
  `allowedValues`

This adds a number of extra errors for some validation that is already
happening for specific options' values, which we'll follow-up and remove
the duplication for.

This also makes sure that we validate this for both repository config
and global config.






## Context

Please select one of the below:

- [ ] This closes an existing Issue: #
- [x] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [x] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [ ] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

The public repository: <URL>

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

jamietanna committed:
fix(config/validation): validate `allowedValues` for arrays

To make sure that we provide faster feedback for users, we should
validate that for options that have an `allowedValues` set, we validate
the option according to the value.

We also need to:

- handle a single option value
- handle an array of option values
- handle when a default (`null`) option is used, as that won't match
  `allowedValues`
- add a special case to /not/ handle the `packageRule.versioning` as it
  may have a `regex:` value which isn't technically in the
  `allowedValues`

This adds a number of extra errors for some validation that is already
happening for specific options' values.

This also makes sure that we validate this for both repository config
and global config.

jamietanna committed:
fix(config/validation): validate `allowedValues` for arrays

To make sure that we provide faster feedback for users, we should
validate that for options that have an `allowedValues` set, we validate
the option according to the value.

We also need to:

- handle a single option value
- handle an array of option values
- handle when a default (`null`) option is used, as that won't match
  `allowedValues`
- handle when a `supportsTemplating` config option is used alongside
  `allowedValues`
- add a special case to handle `packageRule.versioning` differently, as
  it may have a `regex:` value which isn't technically in the
  `allowedValues`

This adds a number of extra errors for some validation that is already
happening for specific options' values, which we'll follow-up and remove
the duplication for.

This also makes sure that we validate this for both repository config
and global config.

RahulGautamSingh commented:
🏎️ PR title udate: `fix(config/validation): validate allowedValues for arrays` .. since this change affect repo/file config validation as well

jamietanna commented:
Hah, I ended up implementing this again today, and then realised "I already have a PR open for that"

jamietanna commented:
The "undesirable" flagged when I did an update to the latest commits on HEAD (some of which have undesirable changes)

jamietanna commented on lib/config/validation.spec.ts:
```diff
@@ -692,9 +692,13 @@ describe('config/validation', () => {
       );
       expect(warnings).toHaveLength(0);
-      expect(errors).toHaveLength(1);
+      expect(errors).toHaveLength(2);
```
In a future change, we could remove the existing validation if we wanted to (albeit I prefer the more specific message `Invalid customType: unknown. Key is not a custom manager`)

viceice commented on lib/config/validation.ts:
```diff
@@ -1002,6 +1004,37 @@ async function validateGlobalConfig(
+      // do nothing
+    } else if (isArray(val, isString)) {
+      const invalidOptions = val.filter(
+        (v) => !option.allowedValues?.includes(v),
```
```suggestion
        (v) => !option.allowedValues!.includes(v),
```
maybe you should store those in a local variable before the filter

viceice reviewed (commented):
we should check all options with allowed values.

if it's versioning it must start with `regex:` or must be in allowed values 


otherwise LGTM

viceice commented:
we need to take into account options which are additionally allow to be regex or glob, like `matchUpdateTypes`

- #40804

viceice commented:
- #40805

jamietanna committed:
test(config/validation): remove `platformConfig` test

As the test case appears to be wrong (we return an error, not a warning)
and doesn't seem to be exercising a field we've had available in recent
Renovate releases, since this test was introduced in
bb6c27faf84022401f8aa4edd3218feadeeed3b4.

jamietanna commented on lib/config/validation.spec.ts:
```diff
@@ -3000,6 +3040,100 @@ describe('config/validation', () => {
+          // @ts-expect-error: contains invalid values
+          allowedUnsafeExecutions: ['some', 'value'],
+
+          // TODO nested
```
TODO

jamietanna reviewed (commented):
Also includes https://github.com/renovatebot/renovate/pull/43832/commits/b324ab78f430953b953062a4a8665f27ebdb417e which didn't seem to be merged correctly in #43832 (maybe I messed up a conflict resolution)

jamietanna committed:
test(config/validation): document existing lack of coverage

So when we change it in the future, we have a failing test.

jamietanna committed:
fix(config/validation): validate children of arrays in global config

As we previously weren't doing so.

jamietanna committed:
fix(types): allow global config to be set in `repositories`

As we appear to document this [0] as behaviour that's possible.

This make sense to be allowed, and to make it so a self-hosted
administrator can set more controlled settings across their configured
repos.

[0]: https://github.com/renovatebot/renovate/blob/6e411730a05a642c22ffa5fa4ab71f0eea4cdad2/lib/config/secrets.spec.ts#L42-L44

viceice commented on lib/config/types.ts:
```diff
@@ -325,7 +325,7 @@ export type UpdateConfig<
 export type RenovateRepository =
   | string
-  | (RenovateConfig & {
+  | (AllConfig & {
```
it seems it needs an update?

secustor commented on lib/config/validation.ts:
```diff
@@ -1169,3 +1186,44 @@ function getPossibleConfigFileNames({
+
+function validateAllowedValues(
+  key: string,
+  val: any,
```
```suggestion
  val: unknown,
```

`any` deactivates typechecking
